### ISSUE TITLE
security.ArrayBound FP caused by the lack of modeling narrowing casts

### ISSUE THREAD

**Issue by zufuliu** [llvm/llvm-project#126884]
It has false positive for following code (online at https://godbolt.org/z/9v8P684rc):
```c++
struct Foo {
    unsigned char get(unsigned char ch) const {
        return s[ch];
    }
    unsigned char s[256];
};

int bar(const Foo &foo, const char *s) {
    int j = 0;
    if (s) {
        const unsigned char ch = s[j];
        j++;
        if (static_cast<signed char>(ch) >= 0) {
            // nop
        } else {
            j += foo.get(ch);
        }
    }
    return j;
}
```

```console
<source>:3:16: warning: Out of bound access to memory preceding the field 's' [clang-analyzer-security.ArrayBound]
    3 |         return s[ch];
      |                ^
[<source>:10:9: note: Assuming 's' is non-null](javascript:;)
   10 |     if (s) {
      |         ^
[<source>:10:5: note: Taking true branch](javascript:;)
   10 |     if (s) {
      |     ^
[<source>:13:13: note: Assuming 'ch' is < 0](javascript:;)
   13 |         if (static_cast<signed char>(ch) >= 0) {
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[<source>:13:9: note: Taking false branch](javascript:;)
   13 |         if (static_cast<signed char>(ch) >= 0) {
      |         ^
[<source>:16:18: note: Calling 'Foo::get'](javascript:;)
   16 |             j += foo.get(ch);
      |                  ^~~~~~~~~~~
[<source>:3:16: note: Access of the field 's' at negative byte offset](javascript:;)
    3 |         return s[ch];
      |                ^~~~~
1 warning generated.
```

**Comment by steakhal** [comment]
I think this one is not the fault of the checker.
We just dont model narrowing integral casts, so it leads us in a wrong state.

I wish I'd know off the top of my head the umbrella ticket for casts. This shoulgd go under that one.

Anyone could link it?

**Comment by steakhal** [comment]
> @steakhal this is probably the umbrella ticket that you're thinking of: https://github.com/llvm/llvm-project/issues/39492

Ah yes. Thanks!

**Comment by zufuliu** [comment]
following is also flagged:
https://godbolt.org/z/Wsc3oKacW
```c++
struct Foo {
    unsigned char get(unsigned char ch) const {
        return s[ch];
    }
    unsigned char s[256];
};

int bar(const Foo &foo, const char *s) {
    int j = 0;
    if (s) {
        const unsigned char ch = s[j];
        j++;
        if (static_cast<signed char>(ch) >= 0) { /* nop */}
        j += foo.get(ch);
    }
    return j;
}
```

```console
<source>:3:16: warning: Out of bound access to memory preceding the field 's' [clang-analyzer-security.ArrayBound]
    3 |         return s[ch];
      |                ^
<source>:10:9: note: Assuming 's' is non-null
   10 |     if (s) {
      |         ^
<source>:10:5: note: Taking true branch
   10 |     if (s) {
      |     ^
<source>:13:13: note: Assuming 'ch' is < 0
   13 |         if (static_cast<signed char>(ch) >= 0) { /* nop */}
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:13:9: note: Taking false branch
   13 |         if (static_cast<signed char>(ch) >= 0) { /* nop */}
      |         ^
<source>:14:14: note: Calling 'Foo::get'
   14 |         j += foo.get(ch);
      |              ^~~~~~~~~~~
<source>:3:16: note: Access of the field 's' at negative byte offset
    3 |         return s[ch];
      |                ^~~~~
1 warning generated.
```

following are not flagged:
https://godbolt.org/z/zMbh1xbh5
```c++
struct Foo {
    unsigned char get(unsigned char ch) const {
        return s[ch];
    }
    unsigned char s[256];
};

int bar(const Foo &foo, const char *s) {
    int j = 0;
    if (s) {
        const unsigned char ch = s[j];
        j++;
        j += foo.get(ch);
    }
    return j;
}
```

https://godbolt.org/z/zb41G9nTE
```c++
int bar(const unsigned char (&foo)[256], const char *s) {
    int j = 0;
    if (s) {
        const unsigned char ch = s[j];
        j++;
        if (static_cast<signed char>(ch) >= 0) {
            // nop
        } else {
            j += foo[ch];
        }
    }
    return j;
}
```

https://godbolt.org/z/KjWaPqohM
```c++
struct Foo {
    unsigned char get(unsigned char ch) const {
        return s[ch];
    }
    unsigned char s[256];
};

int bar(const Foo &foo, const char *s) {
    int j = 0;
    if (s) {
        const unsigned char ch = s[j] & 0xff;
        j++;
        if (static_cast<signed char>(ch) >= 0) {
            // nop
        } else {
            j += foo.get(ch);
        }
    }
    return j;
}
```

**Comment by NagyDonat** [comment]
The (approximately) shortest reproducer of this bug is:
```
int table[256];
int foo(signed char x) {
  unsigned char y = x;
  if (x >= 0)
    return x;
  return table[y];
}
```
I created PR https://github.com/llvm/llvm-project/pull/127062 which adds this (and two related cases) to the tests of the analyzer to highlight and document that this problem exists.

I'll probably create a "duct tape" commit that silences many false positives caused by this bug.

**Comment by zufuliu** [comment]
> I created PR [#127062](https://github.com/llvm/llvm-project/pull/127062) which adds this (and two related cases) to the tests of the analyzer to highlight and document that this problem exists.

Thanks, I filled issue #127188 based on the reduced code.
